File: python/inversions.py
#!/usr/bin/env python3

import sys

inversions = 0

def main():
    global inversions
    arr = []
    fname = sys.argv[1]
    with open(fname, 'r') as f:
        for line in f:
            arr.append(int(line.rstrip('\r\n')))
    count_inv(arr, 0, len(arr)-1)
    print('Number of inversions are:', inversions)
    return

def count_split_inv(arr, start, mid, end):
    global inversions
    # Slice the array from mid
    left = arr[start:mid+1]
    right = arr[mid+1:end+1]
    i = 0
    j = 0
    k = start
    # Sort and merge back the slices, until one exhausts
    while ((i < len(left)) and (j < len(right))):
        if (left[i] <= right[j]):
            arr[k] = left[i]
            i += 1
            k += 1
        else:
            arr[k] = right[j]
            j += 1
            k += 1
            # inversions += elements remaining in left
            inversions += len(left) - i
    # If right exhausted, merge back remaining left
    while (i < len(left)):
        arr[k] = left[i]
        i += 1
        k += 1
    # If left exhausted, merge back remaining right
    while (j < len(right)):
        arr[k] = right[j]
        j += 1
        k += 1
    return

def count_inv(arr, start, end):
    if (start < end):
        mid = int((start + end) / 2)
        count_inv(arr, start, mid)
        count_inv(arr, mid+1, end)
        count_split_inv(arr, start, mid, end)
    return

if (__name__ == '__main__'):
    main()

File: c/inversions.c
#include <stdio.h>
#include <stdlib.h>

#define LLI long long int

LLI inversions = 0;

void countSplitInv(LLI [], LLI, LLI, LLI);
void countInv(LLI[], LLI, LLI);
 
int main() {
    LLI i, size, j, T;
    LLI * A;
    scanf("%lld", &T);
    for (j=0; j<T; j++) {
        scanf("%lld", &size);
        A = malloc(size * sizeof(LLI));
        for (i=0; i<size; i++) {
            scanf("%lld", &A[i]);
        }
        inversions = 0;
        countInv(A, 0, size-1);
        printf("%lld\n", inversions);
        free(A);
    }
    return 0;
}

void countSplitInv(LLI A[], LLI start, LLI mid, LLI end) {
    LLI sizeL = mid - start + 1;
    LLI sizeM = end - mid;
    LLI L[sizeL], M[sizeM];
    LLI i, j, k;
    /* Split the array A[] into L[] and M[] */
    for (i=0; i<sizeL; i++) {
        L[i] = A[start+i];
    }
    for (j=0; j<sizeM; j++) {
        M[j] = A[mid+j+1];
    }
    i = 0;
    j = 0;
    k = start;
    /* Sort the sub-arrays while merging back to A[] */
    while ((i < sizeL) && (j < sizeM)) {
        if (L[i] < M[j]) {
            A[k] = L[i];
            i++;
        }
        else {
            A[k] = M[j];
            j++;
            /* inversions += number of elements remaining in L */ 
            inversions += sizeL - i;
        }
        k++;
    }
    /* If M[] exhausted, put remaining L[] back */
    while (i < sizeL) {
        A[k] = L[i];
        i++;
        k++;
    }
    /* If L[] exhausted, put remaining M[] back */
    while (j < sizeM) {
        A[k] = M[j];
        j++;
        k++;
    }
}

void countInv(LLI A[], LLI start, LLI end) {
    /* Base case is a single element array i.e already sorted */
    if (start < end) {
        LLI mid = (start + end)/2;
        /* Divide and conquer ;) */
        countInv(A, start, mid);
        countInv(A, mid+1, end);
        countSplitInv(A, start, mid, end);
    }
}
